task/control-plane-agent/src/mgs_gimlet.rs

```diff
@@ -235,6 +257,65 @@ impl MgsHandler {
         })
     }
 
+    pub(crate) fn uart_read(
+        &mut self,
+        data: Leased<idol_runtime::W, [u8]>,
+    ) -> Result<usize, ControlPlaneAgentError> {
+        // This function is only called by humility; switch control to it.
+        self.set_uart_client(UartClient::Humility)?;
+
+        let mut i = 0;
+        while i < data.len() {
+            let Some(b) = self.usart.from_rx.pop_front() else {
+                break;
+            };
+
+            if data.write_at(i, b).is_err() {
+                break;
+            }
+
+            i += 1;
+        }
+
+        if !self.usart.from_rx.is_full() {
+            // If `from_rx` was full and our client was set to Humility the last
+            // time we handled a uart interrupt, we disabled the rx interrupt.
+            // Re-enable it now that humility has pulled some data from us and
+            // made space for more.
+            self.usart.usart.enable_rx_interrupt();
+        }
+
+        Ok(i)
+    }
+
+    pub(crate) fn uart_write(
+        &mut self,
+        data: Leased<idol_runtime::R, [u8]>,
+    ) -> Result<usize, ControlPlaneAgentError> {
+        const CHUNK_SIZE: usize = 32;
+
+        // This function is only called by humility; switch control to it.
+        self.set_uart_client(UartClient::Humility)?;
+
+        let mut chunk = [0; CHUNK_SIZE];
+        let mut i = 0;
+
+        while self.usart.tx_buffer_remaining_capacity() > 0 && i < data.len() {
+            // Min of all three: remaining buffer, remaining data, sizeof(chunk)
+            let n = usize::min(
+                self.usart.tx_buffer_remaining_capacity(),
+                usize::min(CHUNK_SIZE, data.len() - i),
+            );
+            if data.read_range(i..i + n, &mut chunk[..n]).is_err() {
+                return Ok(i);
```

Comment:
Why does this return `Ok` instead of `Err`?  I'd expect a lease read to never fail here unless the lender died, which seems like it should be an error.

Reply:
Mmm yeah that's fair, this should really map to `ClientError::WentAway`, I guess? Out of curiosity - who receives that error if the client is gone?

Reply:
Good question!  I'd still like this to return an error instead of `Ok(i)`, but am not exactly sure where that error goes; maybe it's dropped by the kernel?

Reply:
For posterity: in practice there is no difference here; the error / Ok is indeed dropped if the client is gone. However, it's certainly clearer to return an appropriate error; fixed in 24ec457d.